### DISCUSSION TITLE
Widgets in client app

### DISCUSSION THREAD

**Discussion by intellimedhu** [OrchardCMS/OrchardCore#8885] [Q&A]
I'd like to create a VueJs app and want to use widgets. Can I render those through an API controller based on client routes?

**Comment by Skrypt** [comment]
Your question needs some specifications. Widgets from the FlowPart? You want to retrieve / render a specific Widget from an API call?

**Reply by intellimedhu** [reply]
Have some route in Vue app:

localhost/ **\#/**

localhost/ **#/about**

localhost/ **#/services**

---

Create layers in OC admin:
home   -> rule: url("~/#/")

about  -> rule: url("~/#/about")

services -> rule: url("~/#/services")

---

Assign some html widgets to layers above

---

When I navigate to a route in Vue app I'd like to fetch all the widgets assigned to the layer matching the route.

Is it possible somehow?

**Reply by Skrypt** [reply]
You can use the GraphQL endpoint to retrieve these Widgets per rule/zone but you will need to create some javascript business logic code by yourself. Also, I think it won't support Liquid in those Widgets.

**Reply by intellimedhu** [reply]
Javascript is clear but isn't the backend. Can you show me some suggestive LOC how the backend **query** should look like? Thank you!

**Comment by deanmarcussen** [comment]
Layer Widgets are probably not a great choice for the design you are describing.

While @Skrypt is correct, you can retrieve layer widgets via GraphQl, they are designed to provide layers over the top of pages as they are rendered.

I would suggest the better choice would be to add `Page` content items to your site, (which allow you to add widgets to a Page, i.e using the FlowPart) one for each route you want to have.

You can then retrieve these by filtering on the `AutoroutePart` and using the `Render` option to retrieve the html, or the content itself, if you want to render them yourself.

You can use the GraphiQl explorer to build and experiment with these queries.

**Reply by intellimedhu** [reply]
Thank you! I'll try this solution.

**Reply by intellimedhu** [reply]
When using `Render` option the full source of the page is returned.

![image](https://user-images.githubusercontent.com/65296389/111601941-298eb680-87d3-11eb-9a74-c006d704585f.png)

Is it possible to render the widget html only?

**Reply by Skrypt** [reply]
It's not impossible to achieve but it would require some code and some thinking on how we should implement this properly.
First here, you can get the Content of this widget by retrieving the Html or Paragraph part of the Widget. But these won't be parsed by Liquid if you use them as is. So, what you need is a way to render this HtmlBodyPart or ParagraphPart by getting it parsed by our Liquid parser. Or, not use any Liquid in these Widgets for now.

We don't have a `{{ "some liquid" | liquid }}` filter because it would bite it's own tail ...
But, I think you could use this method with a Razor template though : 
https://docs.orchardcore.net/en/dev/docs/reference/modules/Liquid/#liquidtohtmlasync

**Reply by jtkech** [reply]
@Skrypt 

Just for info the `LiquidFilter` => `| liquid` already exists, see in the `OC.Liquid` module

**Comment by intellimedhu** [comment]
Based on these instructions I'll be able to do what I need. Thanks a lot!
